task-maker-async/src/store.rs

```diff
@@ -0,0 +1,114 @@
+
+    /// Appends a chunk to the input file; the InputFileWriteHandle must have been obtained through
+    /// create_input_file. Renews the lease on the corresponding hash.
+    async fn append_input_chunk(file_handle: InputFileWriteHandle, data: Vec<u8>)
+        -> Result<(), ()>;
+
+    /// Finalizes the input file; no more data can be appended, and the resulting hash must match
+    /// the claimed hash when create_input_file was called.
+    async fn finalize_input_write_file(file_handle: InputFileWriteHandle) -> Result<(), ()>;
+
+    /// Refreshes the lease for the given input file, or creates one if none was present.
+    /// It is guaranteed that the input file will not be deleted while there's an outstanding lease
+    /// to it. It is an error to refresh a lease of a non-existent input file.
+    async fn refresh_input_lease(hash: InputFileHash) -> Result<(), ()>;
+
+    /// Refreshes the given computation writing lease. Used by workers while they're running an
+    /// evaluation group.
+    async fn refresh_computation_write_lease(handle: ComputationWriteHandle) -> Result<(), ()>;
+
+    /// Appends data to a file in the output of a computation. The file is created if it does not
+    /// exist. Refreshes the writing lease.
+    async fn append_computation_chunk(
+        handle: ComputationWriteHandle,
+        file: ComputationDataFile,
+    ) -> Result<(), ()>;
+
+    /// Finalizes a computation output. Terminates the writing lease.
+    async fn finalize_computation(handle: ComputationWriteHandle);
+
+    /// Opens an input file for reading. Creates a lease for the input file that will prevent it
+    /// from being dropped. If the file is not present, returns an error immediately.
+    async fn read_input_file(hash: InputFileHash) -> Result<ReadHandle, ()>;
+
+    /// Opens a computed file for reading. Creates a lease for the computation data that will prevent it
+    /// from being dropped. If the file is not present, waits until the file is created.
+    /// In particular, the server will use this method to read the outcome of the computation; when
+    /// it reads a EndOfFile response from read_chunk, it will know that the computation has
+    /// terminated.
+    async fn read_computation_data_file(
+        hash: ComputationHash,
+        file: ComputationDataFile,
+    ) -> Result<ReadHandle, ()>;
+
+    /// Tries to read from a file. Refreshes the corresponding lease.
+    async fn read_chunk(file: ReadHandle) -> FileReadingOutcome;
+
+    /// Refreshes the given reading lease.
+    async fn refresh_reading_lease(file: ReadHandle) -> Result<(), ()>;
+}
+
+pub struct StoreService {
+    // TODO
+}
+
+impl StoreService {
+    /// Creates the storage for a given computation; this method is called by the server to obtain
+    /// a writing handle that the workers can use. It is an error to create a computation if
+    /// another computation with the same hash already exists, even if it is temporary (i.e. not
```

Comment:
Why? What should we do if the previous writer for this computation died?

Reply:
We realize that when the lease expires, and at that point there shouldn't be another such computation, no?

Reply:
Then reading from an in-progress computation gets very messy, because you don't know if you're reading files from the same attempt.

Reply:
How about dropping an in-progress computation invalidates all the handles to it, and this as a consequence makes all attempts to read those computations fail with an error?

Reply:
Let's say that I want to read two files from the same computation. How can I guarantee that they come from the same attempt?

Reply:
Good question. How about getting a handle-to-computation which then can be used to access the files in the computation itself?

Reply:
Yeah, and then we don't really need handles for individual files (though they might still be useful as a way to keep the read interface identical for both; that said, we should consider which handles holds leases and which take which others with them into explicit oblivion).

Reply:
I'd say that it's useful to keep per-file handles - but all the handles related to a computation share the same lease, and they all share the same fate

Reply:
Clarified this.